scripts/zkai_native_attention_mlp_adapter_opening_geometry_budget_gate.py

```
@@ -0,0 +1,885 @@
+        raise AdapterOpeningGeometryBudgetError(f"failed to write {path}: {err}") from err
+
+
+def fsync_parent_dir(path: pathlib.Path) -> None:
+    if os.name == "nt":
+        return
+    dir_fd: int | None = None
+    try:
+        dir_fd = os.open(path.parent, os.O_RDONLY)
+        os.fsync(dir_fd)
+    except OSError as err:
+        raise AdapterOpeningGeometryBudgetError(f"failed to fsync parent directory for {path}: {err}") from err
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)